edb/pgsql/compiler/dml.py

```diff
@@ -839,33 +838,99 @@ def compile_policy_check(
         relctx.include_rvar(ictx.rel, dml_rvar, path_id=subject_id, ctx=ictx)
 
-        cond_ref = clauses.compile_filter_clause(
-            policy_expr.expr, policy_expr.cardinality, ctx=ictx)
+        # split and compile
+        allow, deny = [], []
+        for policy in access_policies.policies:
+            cond_ref = clauses.compile_filter_clause(
+                policy.expr, policy.cardinality, ctx=ictx
+            )
+
+            if policy.action == qltypes.AccessPolicyAction.Allow:
+                allow.append((policy, cond_ref))
+            else:
+                deny.append((policy, cond_ref))
+
+        def raise_if(
+            a: pgast.BaseExpr, b: bool, msg: pgast.BaseExpr
+        ) -> pgast.BaseExpr:
+            return pgast.FuncCall(
+                name=('edgedb', 'raise_on_null'),
+                args=[
+                    pgast.FuncCall(
+                        name=('nullif',),
+                        args=[a, pgast.BooleanConstant(val=str(b))],
+                    ),
+                    pgast.StringConstant(val='insufficient_privilege'),
+                    pgast.NamedFuncArg(
+                        name='msg',
+                        val=msg,
+                    ),
+                    pgast.NamedFuncArg(
+                        name='table',
+                        val=pgast.StringConstant(val=str(typeref.id)),
+                    ),
+                ],
+            )
 
-        op = 'insert' if isinstance(ir_stmt, irast.InsertStmt) else 'update'
+        if isinstance(ir_stmt, irast.InsertStmt):
+            op = 'insert'
+        else:
+            op = 'update'
         msg = f'access policy violation on {op} of {typeref.name_hint}'
-        maybe_raise = pgast.FuncCall(
-            name=('edgedb', 'raise_on_null'),
-            args=[
-                pgast.FuncCall(
-                    name=('nullif',),
-                    args=[cond_ref, pgast.BooleanConstant(val='false')],
-                ),
-                pgast.StringConstant(val='insufficient_privilege'),
-                pgast.NamedFuncArg(
-                    name='msg',
-                    val=pgast.StringConstant(val=msg),
-                ),
-                pgast.NamedFuncArg(
-                    name='table',
-                    val=pgast.StringConstant(val=str(typeref.id)),
-                ),
-            ],
-        )
+
+        # allow
+        if allow:
+            allow_conds = (cond for _, cond in allow)
+            allow_expr = astutils.extend_binop(None, *allow_conds, op='OR')
+        else:
+            allow_expr = pgast.BooleanConstant(val='false')
+
+        allow_msgs = (pol.error_msg for pol, _ in allow if pol.error_msg)
+        if allow_msgs:
+            hint = 'none of these allow policies match: '
+            hint += ', '.join(allow_msgs)
+            allow_msg = f'{msg} ({hint})'
+        else:
+            allow_msg = f'{msg} (no allow policies)'
```

Comment:
I think I'd prefer to drop most of the extra helpful messages here, and just say `access policy violation` and then, if applicable, following it, the relevant specified error messages.

My reasoning here is that the particular error might be sensitive information, and we don't want to give out any more than what the user asked to give. I'm willing to be convinced that this is not necessary, though.

I think it would be best to be able to produce error messages from both allow and deny rules at the same time, too?

The good news is I think all of these nitpicks should simplify the code, not make it more complex :P

Reply:
I agree, given that this is security it should not be verbose when denying access.